### PR TITLE
21522-Better-support-for-write-barrier-in-MirrorPrimitives

### PR DESCRIPTION
**Pull request by dionisiydk**
MirrorModificationForbidden exception is added as subclass of MirrorModificationForbidden.
It overriddes #retryModificaiton method to use original mirror based selector.

In addition this fix makes MoridicationForbidden kind of Error instead of Exception.
Because many tools handle Error instead of general Exception (SUnit for example). And with previous version this failure was skipped by them which is wrong.
And generally this failure is kind of error.

Also better name for instance creation method:
	for: anObject at: fieldIndex with: newValue retrySelector:
	
And more tests to cover new behaviour.

https://pharo.fogbugz.com/f/cases/21522/Better-support-for-write-barrier-in-MirrorPrimitives

### CODE REVIEW

**Comment by dionisiydk** [comment]
I am looking at problem. 
There is ContextTest>>testJumps with very strange logic

File: src/Kernel/Context.class.st [hunk 1]
```diff
@@ -537,7 +537,7 @@ Context >> completeCallee: aContext [
 		whileTrue: [
 			current := context.
 			nextContext := context quickStep.
-			nextContext ifNil: [ self halt ].
+			nextContext ifNil: [ self error: 'Should not happens!' ].
```
Comment (pavel-krivanek) [inline src/Kernel/Context.class.st]: should be "happen"

Comment (jecisc) [reply src/Kernel/Context.class.st]: Maybe a quick explanation of mhy it should not happen can help if one day it happen?

Comment (dionisiydk) [reply src/Kernel/Context.class.st]: Guys, I have no idea why. I just remove halt by error

**Pull request merged**